shared/src/output.rs

```diff
@@ -0,0 +1,365 @@
+    };
+
+    eprintln!("{}", text.to_ansi_string());
+}
+
+pub fn run_command<E, F: FnOnce(std::io::Error) -> E, F2: FnOnce(Output) -> E>(
+    mut command: Command,
+    quiet: bool,
+    io_error_fn: F,
+    exit_status_fn: F2,
+) -> Result<Output, E> {
+    let child = if quiet {
+        command.output_and_write_streams(std::io::sink(), std::io::sink())
+    } else {
+        command.output_and_write_streams(
+            line_mapped(std::io::stdout(), add_prefix_to_non_empty("    ")),
+            line_mapped(std::io::stderr(), add_prefix_to_non_empty("    ")),
+        )
+    };
+
+    child.map_err(io_error_fn).and_then(|output| {
+        if output.status.success() {
+            Ok(output)
+        } else {
+            Err(exit_status_fn(output))
+        }
+    })
+}
+
+fn add_prefix_to_non_empty<P: Into<Vec<u8>>>(prefix: P) -> impl Fn(Vec<u8>) -> Vec<u8> {
+    let prefix = prefix.into();
+
+    move |mut input| {
+        if input.is_empty() {
+            vec![]
+        } else {
+            let mut result = prefix.clone();
+            result.append(&mut input);
+            result
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BuildpackOutputText {
+    pub line_prefix: Option<String>,
+    pub default_code: Option<String>,
+    pub reset_code: String,
+    pub value_code: String,
+    pub sections: Vec<BuildpackOutputTextSection>,
+}
+
+impl Default for BuildpackOutputText {
+    fn default() -> Self {
+        Self {
+            line_prefix: None,
+            default_code: None,
+            reset_code: String::from(ANSI_RESET_CODE),
+            value_code: String::from(ANSI_VALUE_CODE),
+            sections: vec![],
+        }
+    }
+}
+
+impl BuildpackOutputText {
+    pub fn new(sections: impl Into<Vec<BuildpackOutputTextSection>>) -> Self {
+        Self {
+            sections: sections.into(),
+            ..Self::default()
+        }
+    }
+
+    fn to_ansi_string(&self) -> String {
+        let mut result = String::new();
+
+        // Every line must start with a style reset, the default ANSI code and the line prefix if
+        // it exists.
+        let line_start = format!(
+            "{}{}{}",
+            ANSI_RESET_CODE,
+            self.default_code.clone().unwrap_or_default(),
+            self.line_prefix.clone().unwrap_or_default()
+        );
+
+        result.push_str(&line_start);
+
+        for section in &self.sections {
+            let text = match section {
+                BuildpackOutputTextSection::Regular(text)
+                | BuildpackOutputTextSection::Value(text)
+                | BuildpackOutputTextSection::Url(text)
+                | BuildpackOutputTextSection::Command(text) => text,
+            };
+
+            match section {
+                BuildpackOutputTextSection::Regular(_) => {}
+                BuildpackOutputTextSection::Value(_) => {
+                    result.push_str(ANSI_VALUE_CODE);
+                    result.push(VALUE_DELIMITER_CHAR);
```

Comment:
I think the spec calls for "value" types (e.g. `Value` and `Url`) to have color applied only to the value, and not the delimiter chars -- at least that's [how bullet_stream implements the spec](https://github.com/heroku-buildpacks/bullet_stream/blob/ec6a37f84c22d42f9a090bb8fbf45c3367373421/src/style.rs#L20-L21).

Seems like a sensible choice for CLI output?